# Stock Analysis Case 2

## Business Problem

Working with the same dataset as in case 1,
* find all stocks above their 200 day moving average
* find all stocks below their 200 day moving average
* calculate 50 day moving average for each stock and see how it relates to 200 day moving average
* is there correlation between the two and the direction a stock moves?
* choose 6 or 7 stocks that seem interesting to buy and hold for a month  

## Part 1: Processing files: calculations, identifying high and low stocks, writing to dataframes

#### Goal:
* For each stock, calculate running 200 day and 50 day averages.
* Find stocks that are currently above their running 200 day average, and those currently below.  Save into appropriate dataframes.
* Also calculate crosses (points where 50 day average crosses 200 day average), days where the 50 day average is above the 200 day average, and the stock direction for each day.  Write columns for stock direction and whether the 50 day average is above the 200 day average to a dataframe for use in a logistic regression in part 3.


In [1]:
import google.datalab.storage as storage
import pandas as pd
import numpy as np
from io import BytesIO

In [2]:
# setting up to read data from bucket: filelist returns a list with an object for each file in the bucket 

mybucket = storage.Bucket('st0ckbucket')
filelist = list(mybucket.objects())

In [3]:
# create empty dataframes to populate during loop

high_stocks = pd.DataFrame(columns=['Stock', 'Last_Adj_Close_Price', '200d_Running_Avg', '50d_Running_Avg'])
low_stocks = pd.DataFrame(columns=['Stock', 'Last_Adj_Close_Price', '200d_Running_Avg', '50d_Running_Avg'])
even_stocks = pd.DataFrame(columns=['Stock', 'Last_Adj_Close_Price', '200d_Running_Avg', '50d_Running_Avg'])

more_analysis_df = pd.DataFrame(columns=['50_beats_200', 'up'])

In [4]:
# filelist = filelist[7:8] # toggle on to work with only one file

In [5]:
for file in filelist:
  
  # get filename, stock ticker name, open file, read into df
  filename = str(file)[45:]
  data_csv = mybucket.object(filename)
  data = data_csv.read_stream()
  df = pd.read_csv(BytesIO(data))
  ticker = filename[13:-4]
  
  # skip empty files
  isempty = df.empty
  if isempty:
    continue
    
  # skip files with less then 200 rows-- 200 day running average can't be calculated
  if len(df) < 200:
    continue
    
  print("\n\n Currently working on ", ticker) # toggle for logging
  
  # flip the df because it's currently in desc order, we want asc to compute running averages
  df.sort_values(by=['date'], inplace=True)
  
  # compute running averages, add to df
  df['200d_rolling_avg'] = df.adjclose.rolling(200, min_periods=1).mean()
  df['50d_rolling_avg'] = df.adjclose.rolling(50,min_periods=1).mean()

  # compare 50 & 200, find stock direction
  df['50_beats_200'] = df['50d_rolling_avg'] > df['200d_rolling_avg']
  df['cross'] = df['50_beats_200'].shift() != df['50_beats_200']
  df['up'] = df['adjclose'] > df['adjclose'].shift()

  # grab the latest value, grab adjclose and 200d_rolling_avg
  last = df.tail(1)
  last_price = last['adjclose'][0]
  last_average = last['200d_rolling_avg'][0]

  # determine if last value is higher or lower then 200 day running average, write to appropriate dataframe
  if last_price > last_average:
    high_stocks = high_stocks.append({'Stock': ticker, 'Last_Adj_Close_Price': last_price, '200d_Running_Avg': last_average, '50d_Running_Avg': last['50d_rolling_avg'][0]},  ignore_index = True)
  elif last_price < last_average:
    low_stocks = low_stocks.append({'Stock': ticker, 'Last_Adj_Close_Price': last_price, '200d_Running_Avg': last_average, '50d_Running_Avg': last['50d_rolling_avg'][0]},  ignore_index = True)
  else:
    even_stocks = even_stocks.append({'Stock': ticker, 'Last_Adj_Close_Price': last_price, '200d_Running_Avg': last_average, '50d_Running_Avg': last['50d_rolling_avg'][0]},  ignore_index = True)
    
  # write stock direction column & 50 day average > 200 day average to df for use in part 3's linear regression
  new_df = df[['50_beats_200', 'up']]
  more_analysis_df = more_analysis_df.append(new_df)
  
  print("more_analysis_df has length ", len(more_analysis_df)) # toggle on or off to log size of cumulative dataframe
  
  # delete dataframe to ensure each file is treated fresh
  del df



 Currently working on  AAAU
more_analysis_df has length  474


 Currently working on  AACG
more_analysis_df has length  807


 Currently working on  AADR
more_analysis_df has length  3010


 Currently working on  AAL
more_analysis_df has length  6727


 Currently working on  AAMC
more_analysis_df has length  8523


 Currently working on  AAME
more_analysis_df has length  17744


 Currently working on  AAN
more_analysis_df has length  25529


 Currently working on  AAOI
more_analysis_df has length  27224


 Currently working on  AAON
more_analysis_df has length  34137


 Currently working on  AAP
more_analysis_df has length  38817


 Currently working on  AAPL
more_analysis_df has length  48790


 Currently working on  AAT
more_analysis_df has length  51173


 Currently working on  AAU
more_analysis_df has length  54831


 Currently working on  AAWW
more_analysis_df has length  58511


 Currently working on  AAXJ
more_analysis_df has length  61500


 Currently working on  AB
more_anal

more_analysis_df has length  499252


 Currently working on  AGEN
more_analysis_df has length  504385


 Currently working on  AGFS
more_analysis_df has length  505749


 Currently working on  AGGP
more_analysis_df has length  506793


 Currently working on  AGGY
more_analysis_df has length  507964


 Currently working on  AGI
more_analysis_df has length  510807


 Currently working on  AGIO
more_analysis_df has length  512556


 Currently working on  AGLE
more_analysis_df has length  513624


 Currently working on  AGM-A
more_analysis_df has length  517336


 Currently working on  AGM
more_analysis_df has length  523419


 Currently working on  AGMH
more_analysis_df has length  523973


 Currently working on  AGNC
more_analysis_df has length  527028


 Currently working on  AGNCN
more_analysis_df has length  527753


 Currently working on  AGND
more_analysis_df has length  529133


 Currently working on  AGQ
more_analysis_df has length  532047


 Currently working on  AGR
more_analysi



 Currently working on  AMKR
more_analysis_df has length  931333


 Currently working on  AMLP
more_analysis_df has length  933814


 Currently working on  AMN
more_analysis_df has length  938505


 Currently working on  AMNB
more_analysis_df has length  943576


 Currently working on  AMOT
more_analysis_df has length  952542


 Currently working on  AMOV
more_analysis_df has length  957388


 Currently working on  AMP
more_analysis_df has length  961104


 Currently working on  AMPE
more_analysis_df has length  963642


 Currently working on  AMPH
more_analysis_df has length  965159


 Currently working on  AMPY
more_analysis_df has length  967217


 Currently working on  AMRB
more_analysis_df has length  972007


 Currently working on  AMRC
more_analysis_df has length  974512


 Currently working on  AMRH
more_analysis_df has length  975238


 Currently working on  AMRK
more_analysis_df has length  976813


 Currently working on  AMRN
more_analysis_df has length  982808


 Currently



 Currently working on  ARR
more_analysis_df has length  1352520


 Currently working on  ARTL
more_analysis_df has length  1352897


 Currently working on  ARTNA
more_analysis_df has length  1358943


 Currently working on  ARTW
more_analysis_df has length  1366476


 Currently working on  ARVN
more_analysis_df has length  1366920


 Currently working on  ARW
more_analysis_df has length  1377076


 Currently working on  ARWR
more_analysis_df has length  1381160


 Currently working on  ARYA
more_analysis_df has length  1381408


 Currently working on  ARYAU
more_analysis_df has length  1381663


 Currently working on  ASA
more_analysis_df has length  1392382


 Currently working on  ASC
more_analysis_df has length  1394125


 Currently working on  ASEA
more_analysis_df has length  1396435


 Currently working on  ASET
more_analysis_df has length  1397394


 Currently working on  ASFI
more_analysis_df has length  1403249


 Currently working on  ASG
more_analysis_df has length  141190



 Currently working on  AZZ
more_analysis_df has length  1867232


 Currently working on  B
more_analysis_df has length  1877288


 Currently working on  BA
more_analysis_df has length  1890028


 Currently working on  BAB
more_analysis_df has length  1892702


 Currently working on  BABA
more_analysis_df has length  1894159


 Currently working on  BAC
more_analysis_df has length  1906101


 Currently working on  BAF
more_analysis_df has length  1910549


 Currently working on  BAH
more_analysis_df has length  1912970


 Currently working on  BAK
more_analysis_df has length  1914990


 Currently working on  BAL
more_analysis_df has length  1917956


 Currently working on  BAM
more_analysis_df has length  1927109


 Currently working on  BAND
more_analysis_df has length  1927774


 Currently working on  BANF
more_analysis_df has length  1934901


 Currently working on  BANFP
more_analysis_df has length  1937089


 Currently working on  BANR
more_analysis_df has length  1943280


 Curr



 Currently working on  BHE
more_analysis_df has length  2341978


 Currently working on  BHF
more_analysis_df has length  2342715


 Currently working on  BHFAL
more_analysis_df has length  2343168


 Currently working on  BHFAP
more_analysis_df has length  2343496


 Currently working on  BHK
more_analysis_df has length  2348177


 Currently working on  BHLB
more_analysis_df has length  2353206


 Currently working on  BHP
more_analysis_df has length  2363324


 Currently working on  BHR
more_analysis_df has length  2366149


 Currently working on  BHTG
more_analysis_df has length  2367105


 Currently working on  BHV
more_analysis_df has length  2371074


 Currently working on  BHVN
more_analysis_df has length  2371871


 Currently working on  BIBL
more_analysis_df has length  2372536


 Currently working on  BICK
more_analysis_df has length  2374952


 Currently working on  BIF
more_analysis_df has length  2385001


 Currently working on  BIIB
more_analysis_df has length  2392251


more_analysis_df has length  2794271


 Currently working on  BQH
more_analysis_df has length  2798761


 Currently working on  BRBS
more_analysis_df has length  2799329


 Currently working on  BRC
more_analysis_df has length  2807938


 Currently working on  BREW
more_analysis_df has length  2814185


 Currently working on  BRF
more_analysis_df has length  2816989


 Currently working on  BRFS
more_analysis_df has length  2821660


 Currently working on  BRG
more_analysis_df has length  2823238


 Currently working on  BRID
more_analysis_df has length  2830557


 Currently working on  BRKL
more_analysis_df has length  2836163


 Currently working on  BRKR
more_analysis_df has length  2841172


 Currently working on  BRKS
more_analysis_df has length  2847571


 Currently working on  BRN
more_analysis_df has length  2856635


 Currently working on  BRO
more_analysis_df has length  2865017


 Currently working on  BROG
more_analysis_df has length  2865397


 Currently working on  BRPA
m

more_analysis_df has length  3241152


 Currently working on  CAPR
more_analysis_df has length  3243149


 Currently working on  CAR
more_analysis_df has length  3252417


 Currently working on  CARA
more_analysis_df has length  3254034


 Currently working on  CARE
more_analysis_df has length  3257021


 Currently working on  CARG
more_analysis_df has length  3257706


 Currently working on  CARS
more_analysis_df has length  3258483


 Currently working on  CARV
more_analysis_df has length  3262418


 Currently working on  CARZ
more_analysis_df has length  3264628


 Currently working on  CASA
more_analysis_df has length  3265268


 Currently working on  CASH
more_analysis_df has length  3270941


 Currently working on  CASI
more_analysis_df has length  3276993


 Currently working on  CASS
more_analysis_df has length  3282293


 Currently working on  CASY
more_analysis_df has length  3291528


 Currently working on  CAT
more_analysis_df has length  3304268


 Currently working on  CA



 Currently working on  CFMS
more_analysis_df has length  3717811


 Currently working on  CFO
more_analysis_df has length  3719320


 Currently working on  CFR
more_analysis_df has length  3729401


 Currently working on  CFRX
more_analysis_df has length  3730857


 Currently working on  CFX
more_analysis_df has length  3733917


 Currently working on  CFXA
more_analysis_df has length  3734284


 Currently working on  CG
more_analysis_df has length  3736339


 Currently working on  CGA
more_analysis_df has length  3739323


 Currently working on  CGBD
more_analysis_df has length  3740092


 Currently working on  CGC
more_analysis_df has length  3741645


 Currently working on  CGEN
more_analysis_df has length  3746648


 Currently working on  CGIX
more_analysis_df has length  3748452


 Currently working on  CGNX
more_analysis_df has length  3756252


 Currently working on  CGO
more_analysis_df has length  3759938


 Currently working on  CGW
more_analysis_df has length  3763247


 C



 Currently working on  CLRO
more_analysis_df has length  4158642


 Currently working on  CLSD
more_analysis_df has length  4159671


 Currently working on  CLSK
more_analysis_df has length  4160251


 Currently working on  CLTL
more_analysis_df has length  4160849


 Currently working on  CLUB
more_analysis_df has length  4164395


 Currently working on  CLVS
more_analysis_df has length  4166565


 Currently working on  CLWT
more_analysis_df has length  4171645


 Currently working on  CLX
more_analysis_df has length  4183592


 Currently working on  CLXT
more_analysis_df has length  4184342


 Currently working on  CM
more_analysis_df has length  4190035


 Currently working on  CMA
more_analysis_df has length  4200182


 Currently working on  CMBM
more_analysis_df has length  4200440


 Currently working on  CMBS
more_analysis_df has length  4202539


 Currently working on  CMC
more_analysis_df has length  4212679


 Currently working on  CMCL
more_analysis_df has length  4219577


more_analysis_df has length  4678539


 Currently working on  CPTA
more_analysis_df has length  4680244


 Currently working on  CPTAG
more_analysis_df has length  4680967


 Currently working on  CPTAL
more_analysis_df has length  4681702


 Currently working on  CQP
more_analysis_df has length  4685048


 Currently working on  CQQQ
more_analysis_df has length  4687675


 Currently working on  CR
more_analysis_df has length  4697834


 Currently working on  CRAK
more_analysis_df has length  4698994


 Currently working on  CRBN
more_analysis_df has length  4700359


 Currently working on  CRBP
more_analysis_df has length  4701783


 Currently working on  CRC
more_analysis_df has length  4705717


 Currently working on  CREE
more_analysis_df has length  4712608


 Currently working on  CRESY
more_analysis_df has length  4718017


 Currently working on  CRF
more_analysis_df has length  4725091


 Currently working on  CRH
more_analysis_df has length  4731178


 Currently working on  CRH

more_analysis_df has length  5193664


 Currently working on  CWEB
more_analysis_df has length  5194535


 Currently working on  CWEN
more_analysis_df has length  5195828


 Currently working on  CWH
more_analysis_df has length  5196768


 Currently working on  CWI
more_analysis_df has length  5200158


 Currently working on  CWK
more_analysis_df has length  5200643


 Currently working on  CWS
more_analysis_df has length  5201558


 Currently working on  CWST
more_analysis_df has length  5207263


 Currently working on  CWT
more_analysis_df has length  5218865


 Currently working on  CX
more_analysis_df has length  5224099


 Currently working on  CXDC
more_analysis_df has length  5226926


 Currently working on  CXH
more_analysis_df has length  5234760


 Currently working on  CXO
more_analysis_df has length  5238012


 Currently working on  CXP
more_analysis_df has length  5239706


 Currently working on  CXSE
more_analysis_df has length  5241465


 Currently working on  CXW
more_a

more_analysis_df has length  5604163


 Currently working on  DGT
more_analysis_df has length  5608968


 Currently working on  DGX
more_analysis_df has length  5614884


 Currently working on  DGZ
more_analysis_df has length  5617949


 Currently working on  DHC
more_analysis_df has length  5620452


 Currently working on  DHCNI
more_analysis_df has length  5620704


 Currently working on  DHI
more_analysis_df has length  5627764


 Currently working on  DHR
more_analysis_df has length  5638229


 Currently working on  DHT
more_analysis_df has length  5641934


 Currently working on  DHX
more_analysis_df has length  5645183


 Currently working on  DHY
more_analysis_df has length  5650693


 Currently working on  DIA
more_analysis_df has length  5656335


 Currently working on  DIAL
more_analysis_df has length  5656995


 Currently working on  DIAX
more_analysis_df has length  5658399


 Currently working on  DIG
more_analysis_df has length  5661777


 Currently working on  DIM
more_a

more_analysis_df has length  6008746


 Currently working on  DTUL
more_analysis_df has length  6009397


 Currently working on  DTW
more_analysis_df has length  6010237


 Currently working on  DTY
more_analysis_df has length  6011171


 Currently working on  DTYL
more_analysis_df has length  6012148


 Currently working on  DUC
more_analysis_df has length  6019060


 Currently working on  DUK
more_analysis_df has length  6029222


 Currently working on  DUKB
more_analysis_df has length  6029651


 Currently working on  DUKH
more_analysis_df has length  6031530


 Currently working on  DUOT
more_analysis_df has length  6032159


 Currently working on  DURA
more_analysis_df has length  6032577


 Currently working on  DUSA
more_analysis_df has length  6033451


 Currently working on  DUSL
more_analysis_df has length  6034199


 Currently working on  DVA
more_analysis_df has length  6040410


 Currently working on  DVAX
more_analysis_df has length  6044529


 Currently working on  DVD
m

more_analysis_df has length  6397924


 Currently working on  EFC
more_analysis_df has length  6400374


 Currently working on  EFF
more_analysis_df has length  6402142


 Currently working on  EFG
more_analysis_df has length  6405895


 Currently working on  EFL
more_analysis_df has length  6406632


 Currently working on  EFNL
more_analysis_df has length  6408683


 Currently working on  EFO
more_analysis_df has length  6411059


 Currently working on  EFR
more_analysis_df has length  6415238


 Currently working on  EFSC
more_analysis_df has length  6419431


 Currently working on  EFU
more_analysis_df has length  6422419


 Currently working on  EFV
more_analysis_df has length  6426172


 Currently working on  EFX
more_analysis_df has length  6436313


 Currently working on  EFZ
more_analysis_df has length  6439504


 Currently working on  EGAN
more_analysis_df has length  6444353


 Currently working on  EGBN
more_analysis_df has length  6448962


 Currently working on  EGF
more_a

more_analysis_df has length  6816916


 Currently working on  EQS
more_analysis_df has length  6823725


 Currently working on  EQT
more_analysis_df has length  6833886


 Currently working on  EQWL
more_analysis_df has length  6837191


 Currently working on  EQX
more_analysis_df has length  6837593


 Currently working on  ERA
more_analysis_df has length  6839454


 Currently working on  ERC
more_analysis_df has length  6843740


 Currently working on  ERF
more_analysis_df has length  6848669


 Currently working on  ERH
more_analysis_df has length  6852743


 Currently working on  ERI
more_analysis_df has length  6854838


 Currently working on  ERIC
more_analysis_df has length  6864353


 Currently working on  ERII
more_analysis_df has length  6867366


 Currently working on  ERJ
more_analysis_df has length  6872385


 Currently working on  ERM
more_analysis_df has length  6873169


 Currently working on  EROS
more_analysis_df has length  6874839


 Currently working on  ERSX
more_

more_analysis_df has length  7246266


 Currently working on  EWP
more_analysis_df has length  7252271


 Currently working on  EWQ
more_analysis_df has length  7258360


 Currently working on  EWRE
more_analysis_df has length  7259418


 Currently working on  EWS
more_analysis_df has length  7265529


 Currently working on  EWT
more_analysis_df has length  7270549


 Currently working on  EWU
more_analysis_df has length  7276598


 Currently working on  EWUS
more_analysis_df has length  7278528


 Currently working on  EWV
more_analysis_df has length  7281640


 Currently working on  EWW
more_analysis_df has length  7287745


 Currently working on  EWX
more_analysis_df has length  7290791


 Currently working on  EWY
more_analysis_df has length  7295855


 Currently working on  EWZ
more_analysis_df has length  7300846


 Currently working on  EWZS
more_analysis_df has length  7303303


 Currently working on  EXAS
more_analysis_df has length  7308187


 Currently working on  EXC
more_a

more_analysis_df has length  7687109


 Currently working on  FENG
more_analysis_df has length  7689408


 Currently working on  FEO
more_analysis_df has length  7692891


 Currently working on  FEP
more_analysis_df has length  7695003


 Currently working on  FET
more_analysis_df has length  7697073


 Currently working on  FEUL
more_analysis_df has length  7697542


 Currently working on  FEUZ
more_analysis_df has length  7698820


 Currently working on  FEX
more_analysis_df has length  7702097


 Currently working on  FEYE
more_analysis_df has length  7703796


 Currently working on  FEZ
more_analysis_df has length  7708250


 Currently working on  FF
more_analysis_df has length  7710586


 Currently working on  FFA
more_analysis_df has length  7714576


 Currently working on  FFBC
more_analysis_df has length  7722717


 Currently working on  FFBW
more_analysis_df has length  7723330


 Currently working on  FFC
more_analysis_df has length  7727717


 Currently working on  FFEU
more

more_analysis_df has length  7981755


 Currently working on  FLTB
more_analysis_df has length  7983198


 Currently working on  FLTR
more_analysis_df has length  7985460


 Currently working on  FLTW
more_analysis_df has length  7985948


 Currently working on  FLWS
more_analysis_df has length  7991212


 Currently working on  FLXN
more_analysis_df has length  7992812


 Currently working on  FLXS
more_analysis_df has length  8002417


 Currently working on  FLZA
more_analysis_df has length  8002756


 Currently working on  FM
more_analysis_df has length  8004719


 Currently working on  FMAO
more_analysis_df has length  8006986


 Currently working on  FMAT
more_analysis_df has length  8008661


 Currently working on  FMB
more_analysis_df has length  8010149


 Currently working on  FMBI
more_analysis_df has length  8019342


 Currently working on  FMC
more_analysis_df has length  8029504


 Currently working on  FMCI
more_analysis_df has length  8029846


 Currently working on  FMF


more_analysis_df has length  8338723


 Currently working on  FTEK
more_analysis_df has length  8345344


 Currently working on  FTF
more_analysis_df has length  8349572


 Currently working on  FTFT
more_analysis_df has length  8352516


 Currently working on  FTGC
more_analysis_df has length  8354190


 Currently working on  FTHI
more_analysis_df has length  8355670


 Currently working on  FTK
more_analysis_df has length  8359429


 Currently working on  FTLB
more_analysis_df has length  8360773


 Currently working on  FTLS
more_analysis_df has length  8362204


 Currently working on  FTNT
more_analysis_df has length  8364877


 Currently working on  FTRI
more_analysis_df has length  8367387


 Currently working on  FTS
more_analysis_df has length  8369296


 Currently working on  FTSD
more_analysis_df has length  8370648


 Currently working on  FTSI
more_analysis_df has length  8371256


 Currently working on  FTSL
more_analysis_df has length  8373062


 Currently working on  FTS

more_analysis_df has length  8724423


 Currently working on  GENY
more_analysis_df has length  8725259


 Currently working on  GEOS
more_analysis_df has length  8730738


 Currently working on  GER
more_analysis_df has length  8732190


 Currently working on  GERN
more_analysis_df has length  8738210


 Currently working on  GES
more_analysis_df has length  8744226


 Currently working on  GEVO
more_analysis_df has length  8745903


 Currently working on  GF
more_analysis_df has length  8753572


 Currently working on  GFED
more_analysis_df has length  8758635


 Currently working on  GFF
more_analysis_df has length  8770359


 Currently working on  GFI
more_analysis_df has length  8780508


 Currently working on  GFIN
more_analysis_df has length  8781313


 Currently working on  GFNCP
more_analysis_df has length  8782680


 Currently working on  GFY
more_analysis_df has length  8786615


 Currently working on  GGAL
more_analysis_df has length  8791630


 Currently working on  GGB
mo

more_analysis_df has length  9127785


 Currently working on  GPJA
more_analysis_df has length  9128186


 Currently working on  GPK
more_analysis_df has length  9135118


 Currently working on  GPL
more_analysis_df has length  9138646


 Currently working on  GPM
more_analysis_df has length  9142384


 Currently working on  GPMT
more_analysis_df has length  9143146


 Currently working on  GPN
more_analysis_df has length  9148043


 Currently working on  GPOR
more_analysis_df has length  9152523


 Currently working on  GPP
more_analysis_df has length  9154082


 Currently working on  GPRE
more_analysis_df has length  9157672


 Currently working on  GPRK
more_analysis_df has length  9159405


 Currently working on  GPS
more_analysis_df has length  9169550


 Currently working on  GRA
more_analysis_df has length  9179258


 Currently working on  GRAF
more_analysis_df has length  9180067


 Currently working on  GRAM
more_analysis_df has length  9181812


 Currently working on  GRBK
mo

more_analysis_df has length  9540915


 Currently working on  HCA
more_analysis_df has length  9543260


 Currently working on  HCAC
more_analysis_df has length  9543521


 Currently working on  HCAP
more_analysis_df has length  9545325


 Currently working on  HCAT
more_analysis_df has length  9545554


 Currently working on  HCC
more_analysis_df has length  9546356


 Currently working on  HCFT
more_analysis_df has length  9548190


 Currently working on  HCHC
more_analysis_df has length  9550928


 Currently working on  HCI
more_analysis_df has length  9553875


 Currently working on  HCKT
more_analysis_df has length  9559437


 Currently working on  HCM
more_analysis_df has length  9560519


 Currently working on  HCR
more_analysis_df has length  9562501


 Currently working on  HCXY
more_analysis_df has length  9562935


 Currently working on  HCXZ
more_analysis_df has length  9563476


 Currently working on  HD
more_analysis_df has length  9573246


 Currently working on  HDAW
mo

more_analysis_df has length  9982788


 Currently working on  HQI
more_analysis_df has length  9985921


 Currently working on  HQL
more_analysis_df has length  9993017


 Currently working on  HQY
more_analysis_df has length  9994509


 Currently working on  HR
more_analysis_df has length  10001334


 Currently working on  HRB
more_analysis_df has length  10013280


 Currently working on  HRC
more_analysis_df has length  10025334


 Currently working on  HRI
more_analysis_df has length  10028763


 Currently working on  HRL
more_analysis_df has length  10038890


 Currently working on  HROW
more_analysis_df has length  10041225


 Currently working on  HRTG
more_analysis_df has length  10042764


 Currently working on  HRZN
more_analysis_df has length  10045199


 Currently working on  HSBC
more_analysis_df has length  10050475


 Currently working on  HSC
more_analysis_df has length  10060636


 Currently working on  HSDT
more_analysis_df has length  10062013


 Currently working on 

more_analysis_df has length  10370526


 Currently working on  IBMQ
more_analysis_df has length  10370805


 Currently working on  IBN
more_analysis_df has length  10375902


 Currently working on  IBND
more_analysis_df has length  10378438


 Currently working on  IBOC
more_analysis_df has length  10384400


 Currently working on  IBP
more_analysis_df has length  10386008


 Currently working on  IBTX
more_analysis_df has length  10387833


 Currently working on  IBUY
more_analysis_df has length  10388879


 Currently working on  ICAD
more_analysis_df has length  10397775


 Currently working on  ICBK
more_analysis_df has length  10399106


 Currently working on  ICCC
more_analysis_df has length  10406696


 Currently working on  ICCH
more_analysis_df has length  10407251


 Currently working on  ICD
more_analysis_df has length  10408977


 Currently working on  ICF
more_analysis_df has length  10413853


 Currently working on  ICFI
more_analysis_df has length  10417317


 Currently w

more_analysis_df has length  10795293


 Currently working on  IMLP
more_analysis_df has length  10797076


 Currently working on  IMMP
more_analysis_df has length  10799007


 Currently working on  IMMR
more_analysis_df has length  10804199


 Currently working on  IMMU
more_analysis_df has length  10813280


 Currently working on  IMO
more_analysis_df has length  10823441


 Currently working on  IMOM
more_analysis_df has length  10824482


 Currently working on  IMTB
more_analysis_df has length  10825380


 Currently working on  IMTE
more_analysis_df has length  10825990


 Currently working on  IMTM
more_analysis_df has length  10827202


 Currently working on  IMUX
more_analysis_df has length  10828756


 Currently working on  IMV
more_analysis_df has length  10830053


 Currently working on  IMVT
more_analysis_df has length  10830274


 Currently working on  IMXI
more_analysis_df has length  10830978


 Currently working on  INBK
more_analysis_df has length  10833551


 Currently

more_analysis_df has length  11226768


 Currently working on  ITMR
more_analysis_df has length  11227071


 Currently working on  ITOT
more_analysis_df has length  11231208


 Currently working on  ITRI
more_analysis_df has length  11237920


 Currently working on  ITRM
more_analysis_df has length  11238450


 Currently working on  ITRN
more_analysis_df has length  11242166


 Currently working on  ITT
more_analysis_df has length  11248345


 Currently working on  ITUB
more_analysis_df has length  11252967


 Currently working on  ITW
more_analysis_df has length  11264815


 Currently working on  IUS
more_analysis_df has length  11265604


 Currently working on  IUSB
more_analysis_df has length  11267087


 Currently working on  IUSG
more_analysis_df has length  11272070


 Currently working on  IUSS
more_analysis_df has length  11272403


 Currently working on  IVAC
more_analysis_df has length  11278500


 Currently working on  IVAL
more_analysis_df has length  11279871


 Currently 

more_analysis_df has length  11639537


 Currently working on  JJSF
more_analysis_df has length  11648186


 Currently working on  JJT
more_analysis_df has length  11649912


 Currently working on  JJU
more_analysis_df has length  11652188


 Currently working on  JKD
more_analysis_df has length  11656211


 Currently working on  JKE
more_analysis_df has length  11660234


 Currently working on  JKF
more_analysis_df has length  11664257


 Currently working on  JKG
more_analysis_df has length  11668279


 Currently working on  JKH
more_analysis_df has length  11672290


 Currently working on  JKHY
more_analysis_df has length  11680835


 Currently working on  JKI
more_analysis_df has length  11684851


 Currently working on  JKJ
more_analysis_df has length  11688872


 Currently working on  JKK
more_analysis_df has length  11692879


 Currently working on  JKL
more_analysis_df has length  11696893


 Currently working on  JKS
more_analysis_df has length  11699445


 Currently working o

more_analysis_df has length  12047420


 Currently working on  KHC
more_analysis_df has length  12048679


 Currently working on  KIDS
more_analysis_df has length  12049363


 Currently working on  KIE
more_analysis_df has length  12053036


 Currently working on  KIM
more_analysis_df has length  12060242


 Currently working on  KIN
more_analysis_df has length  12061892


 Currently working on  KINS
more_analysis_df has length  12065561


 Currently working on  KIO
more_analysis_df has length  12067299


 Currently working on  KIQ
more_analysis_df has length  12069741


 Currently working on  KIRK
more_analysis_df has length  12074245


 Currently working on  KL
more_analysis_df has length  12078689


 Currently working on  KLAC
more_analysis_df has length  12088695


 Currently working on  KLCD
more_analysis_df has length  12088948


 Currently working on  KLDO
more_analysis_df has length  12089279


 Currently working on  KLDW
more_analysis_df has length  12090501


 Currently worki

more_analysis_df has length  12489157


 Currently working on  LDP
more_analysis_df has length  12491152


 Currently working on  LDRS
more_analysis_df has length  12491783


 Currently working on  LDSF
more_analysis_df has length  12492155


 Currently working on  LE
more_analysis_df has length  12493739


 Currently working on  LEA
more_analysis_df has length  12496419


 Currently working on  LEAD
more_analysis_df has length  12497499


 Currently working on  LEAF
more_analysis_df has length  12499873


 Currently working on  LECO
more_analysis_df has length  12506469


 Currently working on  LEDS
more_analysis_df has length  12508867


 Currently working on  LEE
more_analysis_df has length  12518965


 Currently working on  LEG
more_analysis_df has length  12529100


 Currently working on  LEGH
more_analysis_df has length  12529490


 Currently working on  LEGR
more_analysis_df has length  12530095


 Currently working on  LEJU
more_analysis_df has length  12531659


 Currently wor

more_analysis_df has length  12874680


 Currently working on  LTC
more_analysis_df has length  12881699


 Currently working on  LTHM
more_analysis_df has length  12882133


 Currently working on  LTM
more_analysis_df has length  12887696


 Currently working on  LTRPA
more_analysis_df has length  12889169


 Currently working on  LTRPB
more_analysis_df has length  12889556


 Currently working on  LULU
more_analysis_df has length  12892804


 Currently working on  LUNA
more_analysis_df has length  12896338


 Currently working on  LUV
more_analysis_df has length  12906552


 Currently working on  LVHB
more_analysis_df has length  12907482


 Currently working on  LVHI
more_analysis_df has length  12908437


 Currently working on  LVS
more_analysis_df has length  12912351


 Currently working on  LVUS
more_analysis_df has length  12912670


 Currently working on  LW
more_analysis_df has length  12913586


 Currently working on  LWAY
more_analysis_df has length  12920596


 Currently w

more_analysis_df has length  13385519


 Currently working on  MFD
more_analysis_df has length  13389615


 Currently working on  MFDX
more_analysis_df has length  13390322


 Currently working on  MFEM
more_analysis_df has length  13391028


 Currently working on  MFG
more_analysis_df has length  13394463


 Currently working on  MFGP
more_analysis_df has length  13395178


 Currently working on  MFH
more_analysis_df has length  13396406


 Currently working on  MFIN
more_analysis_df has length  13402473


 Currently working on  MFINL
more_analysis_df has length  13403347


 Currently working on  MFL
more_analysis_df has length  13409067


 Currently working on  MFM
more_analysis_df has length  13417541


 Currently working on  MFMS
more_analysis_df has length  13417958


 Currently working on  MFNC
more_analysis_df has length  13421459


 Currently working on  MFO
more_analysis_df has length  13423474


 Currently working on  MFT
more_analysis_df has length  13430409


 Currently wor



 Currently working on  MOD
more_analysis_df has length  13878477


 Currently working on  MODN
more_analysis_df has length  13880304


 Currently working on  MOG-B
more_analysis_df has length  13886848


 Currently working on  MOGU
more_analysis_df has length  13887244


 Currently working on  MOH
more_analysis_df has length  13891525


 Currently working on  MOM
more_analysis_df has length  13892939


 Currently working on  MOMO
more_analysis_df has length  13894895


 Currently working on  MOO
more_analysis_df has length  13898125


 Currently working on  MOR
more_analysis_df has length  13898681


 Currently working on  MORF
more_analysis_df has length  13898938


 Currently working on  MORN
more_analysis_df has length  13902757


 Currently working on  MORT
more_analysis_df has length  13904991


 Currently working on  MOS
more_analysis_df has length  13913166


 Currently working on  MOSY
more_analysis_df has length  13917674


 Currently working on  MOTI
more_analysis_df has le

more_analysis_df has length  14399517


 Currently working on  MWK
more_analysis_df has length  14400066


 Currently working on  MX
more_analysis_df has length  14402410


 Currently working on  MXC
more_analysis_df has length  14406396


 Currently working on  MXDE
more_analysis_df has length  14406613


 Currently working on  MXDU
more_analysis_df has length  14407043


 Currently working on  MXF
more_analysis_df has length  14416892


 Currently working on  MXI
more_analysis_df has length  14420360


 Currently working on  MXIM
more_analysis_df has length  14428500


 Currently working on  MXL
more_analysis_df has length  14431088


 Currently working on  MYC
more_analysis_df has length  14438206


 Currently working on  MYE
more_analysis_df has length  14447938


 Currently working on  MYGN
more_analysis_df has length  14454166


 Currently working on  MYI
more_analysis_df has length  14461275


 Currently working on  MYJ
more_analysis_df has length  14468337


 Currently working 

more_analysis_df has length  14826200


 Currently working on  NHC
more_analysis_df has length  14834416


 Currently working on  NHF
more_analysis_df has length  14837937


 Currently working on  NHLD
more_analysis_df has length  14842561


 Currently working on  NHTC
more_analysis_df has length  14847774


 Currently working on  NI
more_analysis_df has length  14859721


 Currently working on  NIB
more_analysis_df has length  14862707


 Currently working on  NICE
more_analysis_df has length  14868858


 Currently working on  NID
more_analysis_df has length  14870764


 Currently working on  NIE
more_analysis_df has length  14874128


 Currently working on  NIM
more_analysis_df has length  14881121


 Currently working on  NINE
more_analysis_df has length  14881739


 Currently working on  NIO
more_analysis_df has length  14882194


 Currently working on  NIQ
more_analysis_df has length  14884057


 Currently working on  NJR
more_analysis_df has length  14894156


 Currently working 

more_analysis_df has length  15295438


 Currently working on  NURE
more_analysis_df has length  15296174


 Currently working on  NURO
more_analysis_df has length  15297365


 Currently working on  NUS
more_analysis_df has length  15303306


 Currently working on  NUSA
more_analysis_df has length  15304796


 Currently working on  NUV
more_analysis_df has length  15313125


 Currently working on  NUVA
more_analysis_df has length  15317188


 Currently working on  NUW
more_analysis_df has length  15320047


 Currently working on  NVAX
more_analysis_df has length  15326224


 Currently working on  NVCN
more_analysis_df has length  15327269


 Currently working on  NVCR
more_analysis_df has length  15328456


 Currently working on  NVEC
more_analysis_df has length  15333943


 Currently working on  NVFY
more_analysis_df has length  15336027


 Currently working on  NVG
more_analysis_df has length  15340510


 Currently working on  NVGS
more_analysis_df has length  15342365


 Currently w

more_analysis_df has length  15798166


 Currently working on  ONDK
more_analysis_df has length  15799561


 Currently working on  ONE
more_analysis_df has length  15800132


 Currently working on  ONEO
more_analysis_df has length  15801174


 Currently working on  ONEV
more_analysis_df has length  15802152


 Currently working on  ONEY
more_analysis_df has length  15803003


 Currently working on  ONLN
more_analysis_df has length  15803423


 Currently working on  ONTX
more_analysis_df has length  15805137


 Currently working on  ONVO
more_analysis_df has length  15807247


 Currently working on  OOMA
more_analysis_df has length  15808496


 Currently working on  OPB
more_analysis_df has length  15810037


 Currently working on  OPBK
more_analysis_df has length  15811537


 Currently working on  OPCH
more_analysis_df has length  15817548


 Currently working on  OPER
more_analysis_df has length  15817795


 Currently working on  OPES
more_analysis_df has length  15818086


 Currently

more_analysis_df has length  16141260


 Currently working on  PBP
more_analysis_df has length  16144410


 Currently working on  PBPB
more_analysis_df has length  16146108


 Currently working on  PBR
more_analysis_df has length  16151113


 Currently working on  PBS
more_analysis_df has length  16154892


 Currently working on  PBSM
more_analysis_df has length  16155225


 Currently working on  PBT
more_analysis_df has length  16165196


 Currently working on  PBTP
more_analysis_df has length  16165585


 Currently working on  PBTS
more_analysis_df has length  16165902


 Currently working on  PBUS
more_analysis_df has length  16166281


 Currently working on  PBW
more_analysis_df has length  16170142


 Currently working on  PBY
more_analysis_df has length  16170655


 Currently working on  PBYI
more_analysis_df has length  16172686


 Currently working on  PCB
more_analysis_df has length  16173699


 Currently working on  PCEF
more_analysis_df has length  16176310


 Currently work



 Currently working on  PHM
more_analysis_df has length  16626071


 Currently working on  PHO
more_analysis_df has length  16629739


 Currently working on  PHR
more_analysis_df has length  16629973


 Currently working on  PHUN
more_analysis_df has length  16630614


 Currently working on  PHX
more_analysis_df has length  16636116


 Currently working on  PHYL
more_analysis_df has length  16636423


 Currently working on  PHYS
more_analysis_df has length  16639029


 Currently working on  PI
more_analysis_df has length  16640048


 Currently working on  PIC
more_analysis_df has length  16641505


 Currently working on  PICB
more_analysis_df has length  16644044


 Currently working on  PICK
more_analysis_df has length  16646135


 Currently working on  PICO
more_analysis_df has length  16653230


 Currently working on  PID
more_analysis_df has length  16656955


 Currently working on  PIH
more_analysis_df has length  16658494


 Currently working on  PII
more_analysis_df has length 

more_analysis_df has length  17092271


 Currently working on  PRH
more_analysis_df has length  17094097


 Currently working on  PRI
more_analysis_df has length  17096679


 Currently working on  PRIM
more_analysis_df has length  17099666


 Currently working on  PRK
more_analysis_df has length  17106636


 Currently working on  PRLB
more_analysis_df has length  17108739


 Currently working on  PRMW
more_analysis_df has length  17111167


 Currently working on  PRN
more_analysis_df has length  17114598


 Currently working on  PRNB
more_analysis_df has length  17115051


 Currently working on  PRNT
more_analysis_df has length  17116047


 Currently working on  PRO
more_analysis_df has length  17119305


 Currently working on  PROS
more_analysis_df has length  17119543


 Currently working on  PROV
more_analysis_df has length  17125484


 Currently working on  PRPH
more_analysis_df has length  17131700


 Currently working on  PRPL
more_analysis_df has length  17132551


 Currently wo

more_analysis_df has length  17444779


 Currently working on  PYZ
more_analysis_df has length  17448207


 Currently working on  PZA
more_analysis_df has length  17451406


 Currently working on  PZC
more_analysis_df has length  17455855


 Currently working on  PZD
more_analysis_df has length  17459289


 Currently working on  PZG
more_analysis_df has length  17462823


 Currently working on  PZN
more_analysis_df has length  17466017


 Currently working on  PZT
more_analysis_df has length  17469201


 Currently working on  PZZA
more_analysis_df has length  17476019


 Currently working on  QABA
more_analysis_df has length  17478780


 Currently working on  QADA
more_analysis_df has length  17481182


 Currently working on  QAI
more_analysis_df has length  17484021


 Currently working on  QARP
more_analysis_df has length  17484583


 Currently working on  QAT
more_analysis_df has length  17485990


 Currently working on  QCLN
more_analysis_df has length  17489357


 Currently workin

more_analysis_df has length  17810985


 Currently working on  RDVT
more_analysis_df has length  17811557


 Currently working on  RDVY
more_analysis_df has length  17813167


 Currently working on  RDWR
more_analysis_df has length  17818390


 Currently working on  RDY
more_analysis_df has length  17823227


 Currently working on  RE
more_analysis_df has length  17829458


 Currently working on  REAL
more_analysis_df has length  17829714


 Currently working on  REDU
more_analysis_df has length  17830393


 Currently working on  REED
more_analysis_df has length  17833696


 Currently working on  REET
more_analysis_df has length  17835188


 Currently working on  REG
more_analysis_df has length  17841897


 Currently working on  REGI
more_analysis_df has length  17844025


 Currently working on  REGN
more_analysis_df has length  17851396


 Currently working on  REI
more_analysis_df has length  17853507


 Currently working on  REK
more_analysis_df has length  17856089


 Currently wor



 Currently working on  RODM
more_analysis_df has length  18192763


 Currently working on  ROG
more_analysis_df has length  18202797


 Currently working on  ROIC
more_analysis_df has length  18205481


 Currently working on  ROK
more_analysis_df has length  18215190


 Currently working on  ROKT
more_analysis_df has length  18215607


 Currently working on  ROKU
more_analysis_df has length  18216302


 Currently working on  ROL
more_analysis_df has length  18226464


 Currently working on  ROLL
more_analysis_df has length  18230214


 Currently working on  ROM
more_analysis_df has length  18233578


 Currently working on  ROOF
more_analysis_df has length  18235841


 Currently working on  ROP
more_analysis_df has length  18242988


 Currently working on  RORE
more_analysis_df has length  18243792


 Currently working on  ROSC
more_analysis_df has length  18244115


 Currently working on  ROSE
more_analysis_df has length  18247262


 Currently working on  ROSEU
more_analysis_df has l

more_analysis_df has length  18562452


 Currently working on  SAR
more_analysis_df has length  18565767


 Currently working on  SASR
more_analysis_df has length  18571909


 Currently working on  SATS
more_analysis_df has length  18575057


 Currently working on  SAVA
more_analysis_df has length  18580196


 Currently working on  SAVE
more_analysis_df has length  18582487


 Currently working on  SB
more_analysis_df has length  18585532


 Currently working on  SBAC
more_analysis_df has length  18590819


 Currently working on  SBB
more_analysis_df has length  18593880


 Currently working on  SBBX
more_analysis_df has length  18597790


 Currently working on  SBCF
more_analysis_df has length  18606420


 Currently working on  SBFG
more_analysis_df has length  18611122


 Currently working on  SBGI
more_analysis_df has length  18617435


 Currently working on  SBH
more_analysis_df has length  18620854


 Currently working on  SBI
more_analysis_df has length  18627930


 Currently wor

more_analysis_df has length  18962122


 Currently working on  SGC
more_analysis_df has length  18973544


 Currently working on  SGDJ
more_analysis_df has length  18974869


 Currently working on  SGDM
more_analysis_df has length  18976373


 Currently working on  SGG
more_analysis_df has length  18979371


 Currently working on  SGH
more_analysis_df has length  18980154


 Currently working on  SGLB
more_analysis_df has length  18982407


 Currently working on  SGMA
more_analysis_df has length  18988826


 Currently working on  SGMO
more_analysis_df has length  18993918


 Currently working on  SGMS
more_analysis_df has length  19002608


 Currently working on  SGOC
more_analysis_df has length  19004922


 Currently working on  SGOL
more_analysis_df has length  19007642


 Currently working on  SGRP
more_analysis_df has length  19013018


 Currently working on  SGRY
more_analysis_df has length  19014215


 Currently working on  SGU
more_analysis_df has length  19020393


 Currently w

more_analysis_df has length  19426070


 Currently working on  SMLL
more_analysis_df has length  19426954


 Currently working on  SMLP
more_analysis_df has length  19428897


 Currently working on  SMLV
more_analysis_df has length  19430703


 Currently working on  SMM
more_analysis_df has length  19432742


 Currently working on  SMMD
more_analysis_df has length  19433434


 Currently working on  SMMF
more_analysis_df has length  19437681


 Currently working on  SMMT
more_analysis_df has length  19438996


 Currently working on  SMMV
more_analysis_df has length  19439944


 Currently working on  SMN
more_analysis_df has length  19443251


 Currently working on  SMOG
more_analysis_df has length  19446551


 Currently working on  SMP
more_analysis_df has length  19456706


 Currently working on  SMPL
more_analysis_df has length  19457458


 Currently working on  SMSI
more_analysis_df has length  19463690


 Currently working on  SMTC
more_analysis_df has length  19473584


 Currently 

more_analysis_df has length  19807701


 Currently working on  SPXB
more_analysis_df has length  19808164


 Currently working on  SPXC
more_analysis_df has length  19818322


 Currently working on  SPXE
more_analysis_df has length  19819178


 Currently working on  SPXL
more_analysis_df has length  19822112


 Currently working on  SPXN
more_analysis_df has length  19822548


 Currently working on  SPXS
more_analysis_df has length  19825463


 Currently working on  SPXV
more_analysis_df has length  19825898


 Currently working on  SPXX
more_analysis_df has length  19829565


 Currently working on  SPY
more_analysis_df has length  19836472


 Currently working on  SPYB
more_analysis_df has length  19837616


 Currently working on  SPYD
more_analysis_df has length  19838798


 Currently working on  SPYX
more_analysis_df has length  19839936


 Currently working on  SQ
more_analysis_df has length  19841098


 Currently working on  SQBG
more_analysis_df has length  19843687


 Currently 

more_analysis_df has length  20242621


 Currently working on  SWIR
more_analysis_df has length  20247692


 Currently working on  SWK
more_analysis_df has length  20257854


 Currently working on  SWKH
more_analysis_df has length  20262480


 Currently working on  SWKS
more_analysis_df has length  20271437


 Currently working on  SWM
more_analysis_df has length  20277641


 Currently working on  SWN
more_analysis_df has length  20287779


 Currently working on  SWTX
more_analysis_df has length  20287982


 Currently working on  SWX
more_analysis_df has length  20299921


 Currently working on  SWZ
more_analysis_df has length  20308206


 Currently working on  SXC
more_analysis_df has length  20310459


 Currently working on  SXT
more_analysis_df has length  20320602


 Currently working on  SXTC
more_analysis_df has length  20320979


 Currently working on  SY
more_analysis_df has length  20321277


 Currently working on  SYBT
more_analysis_df has length  20327469


 Currently workin



 Currently working on  TER
more_analysis_df has length  20704556


 Currently working on  TERM
more_analysis_df has length  20705359


 Currently working on  TERP
more_analysis_df has length  20706860


 Currently working on  TESS
more_analysis_df has length  20713280


 Currently working on  TEUM
more_analysis_df has length  20716312


 Currently working on  TEVA
more_analysis_df has length  20725631


 Currently working on  TFC
more_analysis_df has length  20735510


 Currently working on  TFI
more_analysis_df has length  20738734


 Currently working on  TFIV
more_analysis_df has length  20739055


 Currently working on  TFLO
more_analysis_df has length  20740453


 Currently working on  TFSL
more_analysis_df has length  20743767


 Currently working on  TFX
more_analysis_df has length  20753927


 Currently working on  TGA
more_analysis_df has length  20762147


 Currently working on  TGB
more_analysis_df has length  20769251


 Currently working on  TGC
more_analysis_df has leng

more_analysis_df has length  21150243


 Currently working on  TRIB
more_analysis_df has length  21157212


 Currently working on  TRIP
more_analysis_df has length  21159368


 Currently working on  TRMB
more_analysis_df has length  21166915


 Currently working on  TRMD
more_analysis_df has length  21167268


 Currently working on  TRMK
more_analysis_df has length  21176514


 Currently working on  TRMT
more_analysis_df has length  21177210


 Currently working on  TRN
more_analysis_df has length  21189105


 Currently working on  TRND
more_analysis_df has length  21189397


 Currently working on  TRNE
more_analysis_df has length  21189643


 Currently working on  TRNO
more_analysis_df has length  21192260


 Currently working on  TRNS
more_analysis_df has length  21200859


 Currently working on  TROW
more_analysis_df has length  21209478


 Currently working on  TROX
more_analysis_df has length  21211869


 Currently working on  TRP
more_analysis_df has length  21221344


 Currently

more_analysis_df has length  21633416


 Currently working on  UBR
more_analysis_df has length  21635914


 Currently working on  UBS
more_analysis_df has length  21640973


 Currently working on  UBSI
more_analysis_df has length  21648958


 Currently working on  UBT
more_analysis_df has length  21651588


 Currently working on  UBX
more_analysis_df has length  21652225


 Currently working on  UCBI
more_analysis_df has length  21656829


 Currently working on  UCC
more_analysis_df has length  21659911


 Currently working on  UCHF
more_analysis_df has length  21660287


 Currently working on  UCI
more_analysis_df has length  21663257


 Currently working on  UCIB
more_analysis_df has length  21663525


 Currently working on  UCO
more_analysis_df has length  21666443


 Currently working on  UCON
more_analysis_df has length  21666932


 Currently working on  UDN
more_analysis_df has length  21670292


 Currently working on  UDOW
more_analysis_df has length  21672908


 Currently worki

more_analysis_df has length  22103524


 Currently working on  UTL
more_analysis_df has length  22111478


 Currently working on  UTMD
more_analysis_df has length  22116755


 Currently working on  UTRN
more_analysis_df has length  22117191


 Currently working on  UTSI
more_analysis_df has length  22122307


 Currently working on  UTSL
more_analysis_df has length  22123072


 Currently working on  UUP
more_analysis_df has length  22126431


 Currently working on  UUU
more_analysis_df has length  22134382


 Currently working on  UUUU
more_analysis_df has length  22137384


 Currently working on  UVSP
more_analysis_df has length  22142563


 Currently working on  UVV
more_analysis_df has length  22154460


 Currently working on  UVXY
more_analysis_df has length  22156358


 Currently working on  UWM
more_analysis_df has length  22159742


 Currently working on  UXI
more_analysis_df has length  22162969


 Currently working on  UXIN
more_analysis_df has length  22163477


 Currently wor

more_analysis_df has length  22517629


 Currently working on  VMC
more_analysis_df has length  22529491


 Currently working on  VMD
more_analysis_df has length  22529709


 Currently working on  VMI
more_analysis_df has length  22541261


 Currently working on  VMO
more_analysis_df has length  22548335


 Currently working on  VMOT
more_analysis_df has length  22549132


 Currently working on  VMW
more_analysis_df has length  22552376


 Currently working on  VNCE
more_analysis_df has length  22554039


 Currently working on  VNDA
more_analysis_df has length  22557610


 Currently working on  VNE
more_analysis_df has length  22560066


 Currently working on  VNET
more_analysis_df has length  22562381


 Currently working on  VNLA
more_analysis_df has length  22563284


 Currently working on  VNM
more_analysis_df has length  22566016


 Currently working on  VNO
more_analysis_df has length  22575714


 Currently working on  VNOM
more_analysis_df has length  22577236


 Currently worki

more_analysis_df has length  22902339


 Currently working on  WBS
more_analysis_df has length  22910713


 Currently working on  WBT
more_analysis_df has length  22911802


 Currently working on  WCC
more_analysis_df has length  22917123


 Currently working on  WCHN
more_analysis_df has length  22917731


 Currently working on  WCN
more_analysis_df has length  22923294


 Currently working on  WDAY
more_analysis_df has length  22925236


 Currently working on  WDC
more_analysis_df has length  22935741


 Currently working on  WDFC
more_analysis_df has length  22947639


 Currently working on  WDIV
more_analysis_df has length  22949408


 Currently working on  WDR
more_analysis_df has length  22955028


 Currently working on  WEA
more_analysis_df has length  22959628


 Currently working on  WEAT
more_analysis_df has length  22961833


 Currently working on  WEBK
more_analysis_df has length  22963406


 Currently working on  WEC
more_analysis_df has length  22973559


 Currently worki

more_analysis_df has length  23456640


 Currently working on  XAN
more_analysis_df has length  23460263


 Currently working on  XAR
more_analysis_df has length  23462343


 Currently working on  XBI
more_analysis_df has length  23465970


 Currently working on  XBIO
more_analysis_df has length  23466875


 Currently working on  XBIT
more_analysis_df has length  23468190


 Currently working on  XBUY
more_analysis_df has length  23468533


 Currently working on  XCEM
more_analysis_df has length  23469541


 Currently working on  XCOM
more_analysis_df has length  23469829


 Currently working on  XDIV
more_analysis_df has length  23470092


 Currently working on  XEC
more_analysis_df has length  23474566


 Currently working on  XEL
more_analysis_df has length  23486512


 Currently working on  XELA
more_analysis_df has length  23487641


 Currently working on  XELB
more_analysis_df has length  23488964


 Currently working on  XENE
more_analysis_df has length  23490379


 Currently wo

more_analysis_df has length  23791262


 Currently working on  ZIXI
more_analysis_df has length  23798983


 Currently working on  ZKIN
more_analysis_df has length  23799696


 Currently working on  ZLAB
more_analysis_df has length  23800397


 Currently working on  ZM
more_analysis_df has length  23800702


 Currently working on  ZN
more_analysis_df has length  23804101


 Currently working on  ZNH
more_analysis_df has length  23809806


 Currently working on  ZOM
more_analysis_df has length  23810463


 Currently working on  ZROZ
more_analysis_df has length  23813123


 Currently working on  ZS
more_analysis_df has length  23813702


 Currently working on  ZSAN
more_analysis_df has length  23814989


 Currently working on  ZSL
more_analysis_df has length  23817863


 Currently working on  ZTO
more_analysis_df has length  23818789


 Currently working on  ZTR
more_analysis_df has length  23826795


 Currently working on  ZTS
more_analysis_df has length  23828663


 Currently working o

In [6]:
more_analysis_df.head()

50_beats_200     up
473        False  False
472        False  False
471        False   True
470        False   True
469        False   True

In [7]:
len(more_analysis_df)

23835211

In [8]:
high_stocks.head()

Stock  Last_Adj_Close_Price  200d_Running_Avg  50d_Running_Avg
0  AAAU             17.709999         15.873150        17.242400
1  AACG              1.310000          1.265000         0.921600
2  AADR             54.090000         49.414148        48.995000
3  AAMC             17.400000         14.469300        16.487200
4  AAON             53.730000         50.118015        51.466065

In [9]:
low_stocks.head()

Stock  Last_Adj_Close_Price  200d_Running_Avg  50d_Running_Avg
0   AAL                 12.50         21.311617        12.498000
1  AAME                  1.70          2.097400         1.820400
2   AAN                 44.73         48.127072        37.717757
3  AAOI                  9.80         10.227425         9.260900
4   AAT                 28.09         37.968098        27.801254

In [10]:
even_stocks.head()

Empty DataFrame
Columns: [Stock, Last_Adj_Close_Price, 200d_Running_Avg, 50d_Running_Avg]
Index: []

In [11]:
len(high_stocks)

2754

In [12]:
len(low_stocks)

4012

## Part 2: Choosing 6 stocks to buy and hold for a month

We were asked to choose 6 stocks to buy and hold for a month.  We decided to look at the stocks that are currently beneath their 200 day average-- meaning they are low compared to their historical value-- and above their 50 day running average-- meaning that they are high compared to their values over the last 50 days.

In [13]:
# look at the dataframe we already produced in our loop for stocks that are historically low
low_stocks.head()

Stock  Last_Adj_Close_Price  200d_Running_Avg  50d_Running_Avg
0   AAL                 12.50         21.311617        12.498000
1  AAME                  1.70          2.097400         1.820400
2   AAN                 44.73         48.127072        37.717757
3  AAOI                  9.80         10.227425         9.260900
4   AAT                 28.09         37.968098        27.801254

In [14]:
# add a new column capturing whether these stocks are higher then their 50 day running average
low_stocks['Above_50d_avg'] = low_stocks['Last_Adj_Close_Price'] > low_stocks['50d_Running_Avg']

In [15]:
# sort by this new column and view the results
low_stocks.sort_values(by=['Above_50d_avg'], inplace=True, ascending=False)
low_stocks.head(10)

Stock  Last_Adj_Close_Price  200d_Running_Avg  50d_Running_Avg  \
0      AAL             12.500000         21.311617        12.498000   
2451   NJV             12.640000         13.201838        12.554954   
2322  MRIN              1.620000          1.663550         1.499200   
2323   MRK             78.779999         81.282633        78.038632   
2326   MRO              6.120000          9.125400         6.012800   
2328   MSB             17.400000         18.455979        16.377671   
2329  MSBF             12.000000         14.314313        11.712560   
2332   MSD              8.540000          8.684368         8.155791   
2337  MSON             13.440000         15.650800        12.431200   
2341    MT             10.940000         13.476150        10.287400   

      Above_50d_avg  
0              True  
2451           True  
2322           True  
2323           True  
2326           True  
2328           True  
2329           True  
2332           True  
2337           True  
2341           True

We decided to choose six stocks that are currently beneath their 200 day running average-- meaning they are historically low-- but above the 50 day running average meaning that they are high compared to the last 50 days.  Therefore, we chose VMC, FID, FGD, EML,FHI and FHK.

## Part 3: Identifying the relationship between 50 day and 200 day running averages, looking for connections with stock direction

In the loop we created a dataframe with two categorical variables: 
* 50_beats_200 is the relationship between the two running averages, True if the 50 day average is higher
* up is the direction of the stock, True if the stock has increased value since the day before

We will run a logistic regression attempting to predict stock direction (up) based on the predictor value 50_beats_20.

In [16]:
# view the dataframe produced in our loop 
# 50_beats_200 is True if the 50 day running average is above the 200 day running average
# Up is true if the stock is higher then it was yesterday.

more_analysis_df.head()

50_beats_200     up
473        False  False
472        False  False
471        False   True
470        False   True
469        False   True

In [17]:
len(more_analysis_df)
# check the length of the dataframe; we have quite a large dataset

23835211

In [18]:
# prepare the data for logistic regression
# need reshape because we only have one predictor value, logistic regression generally expects more then one

X = more_analysis_df['50_beats_200'].reshape(-1, 1)
y = more_analysis_df['up']

/usr/local/envs/py3env/lib/python3.5/site-packages/ipykernel/__main__.py:4: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead


In [19]:
# needs additional processing to prepare for logistic regression
y = np.array(y.values.tolist())

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.model_selection import train_test_split

In [21]:
# create the model
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
logreg = LogisticRegression()

In [22]:
# fit the model
logreg.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [23]:
# run prediction sand evaluate accuracy of the model
y_pred = logreg.predict(X_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(X_test, y_test)))

Accuracy of logistic regression classifier on test set: 0.53


Based on the low accuracy of the logistic regression (0.53), this is not a good way to predict direction of stock by whether or not the 50 day average is above the 200 day average.

## Next steps:
I suspect that the reason our logistic regression is not successful is that our data is too noisy.  By calculating stock direction based on a single day-- whether one day's value is higher then the previous day-- I have probably introduced too much noise to find correlation.

Next steps would be to smooth the data.  I would first attempt to calculate stock direction looking at a larger number of dates-- perhaps using a running average, perhaps using more manual calculations.  Determining the factor to use for smoothing would be tricky-- should I look at average over a week?  A month?  For this decision,working with several individual files and plotting graphs would probably help determine the right scale of smoothing.

I'd be interested to see if it is possible to create a more accurate logistic regression by filtering out daily fluctuations in this way.